### PR TITLE
Completions: add support for doas as for sudo

### LINKED ISSUE #2047: Autocomplete does not work with sudo
For example when I want to autocomplete nautilus file manager,
```sh
$ sudo naut #press tab here
```
If I press tab to autocomplete `naut` to `nautilus` it won't work. But autocomplete without `sudo` works fine

### PR DESCRIPTION
**Pull request by ZerdoX-x**
# Description

Fixes #2047 but for the `doas` command the same way as in #8094
<details>
<summary>Screenshots</summary>
Fixes this difference:

image
image

</details>

# User-Facing Changes
No breaking changes. If people not using `doas`, no difference at all.

# Tests
I have not added any tests since its using same logic as for "sudo". I guess if something would go wrong in this part, sudo tests will cover it?

# Additional context
As a nushell user I could not find a way to implement custom completion for a "sudo like command". Since I can see `sudo` being hardcoded in sources, this is what I propose.

~~Also I have almost zero knowledge of rust and this is definitely not the clean way yet~~

### CODE REVIEW

File: crates/nu-cli/src/completions/completer.rs [hunk 1]
```diff
@@ -137,6 +137,7 @@ impl NuCompleter {
                             let is_passthrough_command = spans
                                 .first()
                                 .filter(|content| *content == &String::from("sudo"))
+                                .filter(|content| *content == &String::from("doas"))
```
Comment (sholderbach) [inline crates/nu-cli/src/completions/completer.rs:140]: Filtering twice with different predicates would return an empty iterator.
So let's combine them with an or (`||`)

```suggestion
                                .filter(|content| *content == &String::from("sudo") || *content == &String::from("doas"))
```

While we are at it we could make the types a bit nicer (by dereferencing `String` to `&str`)

```suggestion
                                .filter(|content| content.deref() == "sudo" || content.deref() == "doas")
```

**Review by sholderbach — CHANGES_REQUESTED** [review]
Thanks for giving this a shot!
No fault for not being familiar with Rust yet. In this case your changes not yet add support for command completion after `sudo`/`doas`. In the screenshot you can see that it only suggests files (which I think it will fallback to quite liberally)

**Comment by ZerdoX-x** [comment]
@sholderbach thank you so much for review <3

Screenshots are just illustration of the problem, not it's solution :D

I've made changes according to your code advice.

**Comment by ZerdoX-x** [comment]
@sholderbach Hi :)
I've resolved merge conflict and added `doas` to typos false positives so CI check would not complain.
Could you please take a look?

File: crates/nu-cli/src/completions/completer.rs [hunk 2]
```diff
@@ -10,6 +10,7 @@ use nu_protocol::{
     BlockId, PipelineData, Span, Value,
 };
 use reedline::{Completer as ReedlineCompleter, Suggestion};
+use std::ops::Deref;
```
Comment (sholderbach) [inline crates/nu-cli/src/completions/completer.rs:13]: Not needed anymore

```suggestion
```

**Review by sholderbach — APPROVED** [review]
Thanks for iterating on this!

I think we should in the future have a way to simply provide a list of commands or a signature that will take external commands and arguments in the trailing position.

But to get this resolved I am fine with landing this right now so we can unblock

**Comment by fdncred** [comment]
Darn, I forgot about gsudo, that I sometimes use on windows. Now that I think about it, maybe it also has a sudo version of the command?

File: crates/nu-cli/src/completions/completer.rs (hunk 2 above)
Comment (ZerdoX-x) [reply crates/nu-cli/src/completions/completer.rs]: Oh, yeah. Thanks for this

**Comment by ZerdoX-x** [comment]
> I think we should in the future have a way to simply provide a list of commands or a signature that will take external commands and arguments in the trailing position.

Definitely. As Darren mentioned, it may be some other commands or just any command that wraps other binaries. I could not find a way to implement custom completion with current level of interfaces and/or documentation.

Do you want me to open an issue for this?

**Comment by fdncred** [comment]
> Do you want me to open an issue for this?

I think so. Thanks!

**Comment by p00f** [comment]
@ZerdoX-x I used to do it with
```
let doas_completer = {|spans: list<string>|
    do $env.config.completions.external.completer ($spans | skip 1)
}

# This completer will use carapace by default
let external_completer = {|spans|
    let expanded_alias = (scope aliases | where name == $spans.0 | get -i 0.expansion)
    let spans = if $expanded_alias != null  {
        $spans | skip 1 | prepend ($expanded_alias | split words)
    } else {
        $spans
    }

    match $spans.0 {
        doas => $doas_completer
        _ => $carapace_fish_completer
    } | do $in $spans
}
$env.config = {
    completions: {
        external: {
            enable: true
            completer: $external_completer
        }
    }
}
```

**Comment by p00f** [comment]
note that `pkexec` is another sudo-like command

**Comment by ZerdoX-x** [comment]
> before this PR, I used to do it with

I think anyways I'll open another issue since there are plenty of commands who accept another program as their first parameter. The most basic example here is `which`, which is not supported by nushell built-in completions. I am not sure this is user friendly, to make people write this in order to support `which`, `exec` and other commands :D

**Comment by p00f** [comment]
`which` just takes other programs, you don't have to provide completions for the other program's arguments. agree on `exec`

**Pull request merged**